include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,934 @@
+
+        bool appendToList(CCommandSegment::Iterator& listHeadItr, CCommandSegment*& listTail)
+        {
+            auto segment = m_pool.emplace<CCommandSegment>();
+            if (!segment)
+            {
+                assert(false);
+                return false;
+            }
+
+            if (listTail)
+            {
+                listTail->setNext(segment);
+                listTail = listTail->getNext();
+            }
+            else
+            {
+                assert(!listHeadItr.cmd && !listHeadItr.segment && "List should've been empty.");
+
+                listTail = segment;
+                listHeadItr = segment->begin();
+            }
+
+            return bool(segment);
+        }
+
+        void deleteList(CCommandSegment::Iterator& headItr, CCommandSegment*& tail)
+        {
+            if (tail)
+            {
+                assert(headItr.segment && headItr.cmd);
+
+                auto freeSegment = [this](CCommandSegment* segment)
+                {
+                    if (segment)
+                    {
+                        segment->~CCommandSegment();
+                        m_pool.deallocate(segment, COMMAND_SEGMENT_SIZE);
+                    }
+                };
+
+                CCommandSegment* prevSegment = nullptr;
+                for (auto& itr = headItr; itr != tail->end();)
+                {
+                    prevSegment = itr.segment;
+                    auto prevCmd = *(itr++);
+
+                    prevCmd->~ICommand();
+
+                    // No need to deallocate prevCmd because:
+                    // - it has been allocated from the LinearAddressAllocator where deallocate is a No-OP
+                    // - the memory will get reclaimed in `~LinearAddressAllocator` when whole segment is destroyed
+
+                    if (itr.segment == prevSegment)
+                        continue;
+
+                    // Segment has changed, we need to free the previous one.
+                    freeSegment(prevSegment);
+                }
+
+                // The last segment is always freed outside the loop.
+                assert(tail == prevSegment);
+                freeSegment(tail);
+
+                headItr.cmd = nullptr;
+                headItr.segment = nullptr;
+                tail = nullptr;
+            }
+            else
+            {
+                assert(!headItr.segment && !headItr.cmd);
+            }
+        }
+
+        void clear(CCommandSegment*& head, CCommandSegment*& tail)
+        {
+            // The way deleteList is implemented, we also need to keep the tail pointer of each child command segment list.
+            // See my comment for one possible way to solve this:
+            // https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#issuecomment-1345979974
+
+            // for (auto& segment = head; segment; segment = segment->getNextHead())
+            //     deleteList({}, );
+        }
```

Comment:
yeah your proposal is fine, I'd just not keep the tail or need the `dummy`

Reply:
Resolved.